src/main/java/edu/cornell/kfs/tax/batch/service/impl/TaxFileRowWriterBase.java

```diff
@@ -0,0 +1,117 @@
+package edu.cornell.kfs.tax.batch.service.impl;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.Validate;
+import org.kuali.kfs.sys.KFSConstants;
+import org.springframework.beans.BeanWrapper;
+import org.springframework.beans.PropertyAccessorFactory;
+
+import edu.cornell.kfs.tax.batch.CUTaxBatchConstants.TaxOutputFieldType;
+import edu.cornell.kfs.tax.batch.dataaccess.TaxDtoFieldEnum;
+import edu.cornell.kfs.tax.batch.service.TaxFileRowWriter;
+import edu.cornell.kfs.tax.batch.xml.TaxOutputDefinitionV2;
+import edu.cornell.kfs.tax.batch.xml.TaxOutputFieldV2;
+import edu.cornell.kfs.tax.batch.xml.TaxOutputSectionV2;
+
+public abstract class TaxFileRowWriterBase<T> implements TaxFileRowWriter<T> {
+
+    protected final TaxOutputDefinitionV2 taxOutputDefinition;
+    protected final Map<String, TaxOutputSectionV2> taxOutputSections;
+    protected final Map<String, TaxDtoFieldEnum> validDerivedFields;
+    protected final Map<TaxDtoFieldEnum, BiFunction<String, Object, String>> formatters;
+    protected final BiFunction<String, Object, String> defaultFormatter;
+
+    protected TaxFileRowWriterBase(final TaxOutputDefinitionV2 taxOutputDefinition,
+            final Class<? extends TaxDtoFieldEnum> fieldEnumClass) {
+        Validate.notNull(taxOutputDefinition, "taxOutputDefinition cannot be null");
+        Validate.notNull(fieldEnumClass, "fieldEnumClass cannot be null");
+        Validate.isTrue(fieldEnumClass.getEnumConstants() != null, "fieldEnumClass should have been an enum class");
+        this.taxOutputDefinition = taxOutputDefinition;
+        this.taxOutputSections = taxOutputDefinition.getSections().stream()
+                .collect(Collectors.toUnmodifiableMap(TaxOutputSectionV2::getName, Function.identity()));
+        this.validDerivedFields = Arrays.stream(fieldEnumClass.getEnumConstants())
+                .collect(Collectors.toUnmodifiableMap(TaxDtoFieldEnum::getFieldName, Function.identity()));
+        this.formatters = buildFormattersMap();
+        this.defaultFormatter = this::performDefaultFormatting;
+    }
+
+    protected abstract Map<TaxDtoFieldEnum, BiFunction<String, Object, String>> buildFormattersMap();
+
+    protected String performDefaultFormatting(final String sectionName, final Object value) {
+        return value != null ? value.toString() : KFSConstants.EMPTY_STRING;
+    }
+
+    @Override
+    public void writeHeaderRow(final String sectionName) throws IOException {
+        Validate.notBlank(sectionName, "sectionName cannot be blank");
+
+        final TaxOutputSectionV2 section = getSection(sectionName);
+        Validate.validState(section.isHasHeaderRow(),
```

Comment:
Could this be hasHeaderRow() instead of isHasHeaderRow()? For Sprintax won't all of the files have header rows?

Reply:
This is a boolean property, so its getter method needs to start with "is" or "get" to conform to the JavaBean conventions. I agree that all of the current use cases will have this property set to true; however, I'm not certain whether that will be the case on the 1099 side.

In the meantime, though, I could remove this property, and then we can add it back in later if necessary.